src/object_store/src/object/s3.rs

```diff
@@ -12,22 +12,259 @@
+        // empty.
+        //
+        // Reference: <https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html>
+        self.client
+            .abort_multipart_upload()
+            .bucket(&self.bucket)
+            .key(&self.key)
+            .upload_id(&self.upload_id)
+            .send()
+            .await?;
+        Ok(())
+    }
+}
 
-use super::{BlockLocation, ObjectError, ObjectMetadata};
-use crate::object::{Bytes, ObjectResult, ObjectStore};
+#[async_trait::async_trait]
+impl StreamingUploader for S3StreamingUploader {
+    fn write_bytes(&mut self, data: Bytes) -> ObjectResult<()> {
+        fail_point!("s3_write_bytes_err", |_| Err(ObjectError::internal(
+            "s3 write bytes error"
+        )));
+        let data_len = data.len();
+        self.not_uploaded_len += data_len;
+        self.buf.push(data);
+
+        if self.not_uploaded_len > PART_SIZE {
+            if self.part_end == 0 {
+                // Mark current slice of buffer to be the next part to be uploaded.
+                self.part_end = self.buf.len();
+            } else {
+                // `data` should be uploaded in the next part.
+                self.next_part_len += data_len;
+            }
+        }
+        if self.next_part_len >= MIN_PART_SIZE {
+            // Take a 16MiB part and upload it. `Bytes` performs shallow clone.
+            let part = self.buf.drain(..self.part_end).collect();
+            self.upload_part(part, self.not_uploaded_len - self.next_part_len);
+            self.not_uploaded_len = self.next_part_len;
+            self.part_end = 0;
+            self.next_part_len = 0;
+        }
+        Ok(())
+    }
+
+    /// If the data in the buffer is smaller than `MIN_PART_SIZE`, abort multipart upload
+    /// and use `PUT` to upload the data. Otherwise flush the remaining data of the buffer
+    /// to S3 as a new part. Fallback to `PUT` on failure.
+    async fn finish(mut self) -> ObjectResult<()> {
+        fail_point!("s3_finish_streaming_upload_err", |_| Err(
+            ObjectError::internal("s3 finish streaming upload error")
+        ));
+        // Fallback to `PUT`.
+        if self.join_handles.is_empty() {
+            self.abort().await?;
+            self.client
+                .put_object()
+                .bucket(&self.bucket)
+                .key(&self.key)
+                .body(get_upload_body(self.buf))
+                .content_length(self.not_uploaded_len as i64)
+                .send()
+                .await?;
+            return Ok(());
+        }
+        if self.flush_and_complete().await.is_err() {
+            self.abort().await?;
+        }
+        Ok(())
+    }
+}
+
+fn get_upload_body(data: Vec<Bytes>) -> aws_sdk_s3::types::ByteStream {
+    Body::wrap_stream(stream::iter(data.into_iter().map(ObjectResult::Ok))).into()
+}
 
 /// Object store with S3 backend
 pub struct S3ObjectStore {
-    client: Client,
+    client: Arc<Client>,
```

Comment:
I think we don't need to add `Arc` to Client since it is already cloneable:
```
// aws-sdk implementation:
pub struct Client {
    handle: std::sync::Arc<Handle>,
}

impl std::clone::Clone for Client {
    fn clone(&self) -> Self {
        Self {
            handle: self.handle.clone(),
        }
    }
}
```